SJRK-402, SJRK-403 & SJRK-404: Add Signup, Login and Logout capabilities to the tool

BlueSlug opened the pull request:
This update will add Signup, Login & Logout controls and pages to the tool. This is currently only implemented for the `base` and `aihec` themes.

This update also includes a few bug fixes for issues that did not have Jiras filed, namely:
- the `author` data was missing on server responses for the View page
- the `createNewStoryOnServer` function was being called at page load, rather than at first user interaction as we had assumed
- (housekeeping) all `.handlebars` files have been renamed to `.hbs` to reduce filename length for better readability

Jira links:
https://issues.fluidproject.org/browse/SJRK-402
https://issues.fluidproject.org/browse/SJRK-403
https://issues.fluidproject.org/browse/SJRK-404
https://issues.fluidproject.org/browse/SJRK-442


@cindyli to review

**UPDATE:**
This update includes commented-out code which introduces the regression described by [SJRK-454](https://issues.fluidproject.org/browse/SJRK-454)

BlueSlug committed:
SJRK-442: rename `.handlebars` files to `.hbs` for tidy file tree & refs

BlueSlug committed:
SJRK-404 & SJRK-403: add initial author controls UI template

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
SJRK-404 & SJRK-403: add author controls UI, messages, wiring & styling

BlueSlug committed:
SJRK-404: move authorControls to its own grade, add ref to the new file

BlueSlug committed:
SJRK-404: add authorControls UI tests

BlueSlug committed:
SJRK-404: add wiring in Page grade for Log Out button

BlueSlug committed:
SJRK-403: add login page & UI grades and template

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
SJRK-403: add Login form and wire it up

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
SJRK-403/404: add wiring for login/logout server calls, author greeting

BlueSlug committed:
SJRK-403: redirect user to previous page after logging in

BlueSlug committed:
SJRK-403: add login progress indicator and server response text area

BlueSlug committed:
SJRK-404: fix log out button test

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
NOJIRA: fix typo in parameter name (was "visibility", should be "value")

BlueSlug committed:
SJRK-403/404: update server login/logout URLs after endpoints changed

BlueSlug committed:
SJRK-403: add loginUi tests

BlueSlug committed:
SJRK-403: add initial login page tests, more to come

BlueSlug committed:
SJRK-403: add mock server wait option, add more login visibility tests

BlueSlug committed:
SJRK-403: add further login element visibility tests & supporting wiring

BlueSlug committed:
SJRK-403: make login error messages a bit friendlier (still needs work)

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
SJRK-403: add more login page tests

BlueSlug committed:
SJRK-403: add further login page tests for invokers/functions

jobara committed:
SJRK-403: Add JSON Schema validation for login model.

BlueSlug committed:
Merge pull request #34 from jobara/SJRK-404

SJRK-403: Add JSON Schema validation for login model.

BlueSlug committed:
SJRK-404: add logout interface tests (race condition present)

BlueSlug committed:
SJRK-404: fix for page logout tests

BlueSlug committed:
Merge remote-tracking branch 'jobara/SJRK-405' into SJRK-404

BlueSlug committed:
Merge remote-tracking branch 'upstream/main' into SJRK-404

BlueSlug committed:
SJRK-404: bump version for 0.5.0 release

jobara committed:
Merge branch 'main' into SJRK-404

jobara committed:
SJRK-404: Fix broken tests after update from main

BlueSlug committed:
Merge pull request #35 from jobara/SJRK-404

SJRK-404: Update from main

BlueSlug committed:
SJRK-403: add login page to AIHEC theme

BlueSlug committed:
NOJIRA: make page heading mobile styling update

jobara committed:
SJRK-403: Perform local validation before sending to request to server

jobara committed:
SJRK-403: Fix/Update tests for login

BlueSlug committed:
SJRK-404: split authorControls UI into a mixin grade for AIHEC site only

BlueSlug committed:
SJRK-404: split page with authorControls tests into its own sequence

BlueSlug committed:
Merge pull request #36 from jobara/SJRK-404

SJRK-403: Perform local validation before sending to request to server

jobara committed:
SJRK-403: Update to latest Infusion dev release to get fix for FLUID-6597

BlueSlug committed:
Merge pull request #37 from jobara/SJRK-404

SJRK-403: Update to latest Infusion dev release to get fix for FLUID-6597

BlueSlug committed:
SJRK-403: add basic localized login validation and server error messages

jobara committed:
SJRK-403: Updated dependencies

fluid-express-user and other dependencies have been updated to work with the 2.x line of Kettle. I've updated kettle and related dependencies. Also updated other dependencies that have been upated.

BlueSlug committed:
Merge pull request #38 from jobara/SJRK-404

SJRK-403: Updated dependencies

BlueSlug committed:
SJRK-403: have login form submit on enter with focus on text inputs

BlueSlug committed:
SJRK-403: add localized login error messages by field (val & server)

jobara committed:
SJRK-403: Only create a new story on the server after the story's model has been updated.

BlueSlug committed:
Merge pull request #39 from jobara/SJRK-404

SJRK-403: Only create a new story on the server after the story's model has been updated.

BlueSlug committed:
NOJIRA: fix bug that excludes the author from story data on View page

BlueSlug committed:
SJRK-403: fix vertical alignment of author greeting in authorControls

BlueSlug committed:
SJRK-403: solve issues with validation error messages

BlueSlug committed:
NOJIRA: fix server tests after solving missing "author" regression

BlueSlug committed:
SJRK-403: fix issues with login page tests and validation error messages

BlueSlug committed:
SJRK-403: add stub to prevent repeated failing server calls in tests

cindyli commented on src/ui/base-page-login.js:
```diff
@@ -0,0 +1,371 @@
+            passwordErrorMessage: ""
+        },
+        modelListeners: {
+            progressAreaVisible: {
```
Recently, I learnt a a simpler way to control the visibility of a dom element with the new renderer via "modelRelay" block:

```
modelRelay: {
    progressAreaVisibleChange: {
        source: "progressAreaVisible",
        target: "{loginUi}.dom.progressArea.visible"
    }
}
```

Just for your information. You don't need to make the change.

jobara replied:
@cindyli thanks for this info. We've updated to the FLUID-6580 line of Infusion so I guess this should be do-able now. @BlueSlug, if you decide to do this change later, could you please file a JIRA about it.

BlueSlug replied:
I knew this was on its way, and it looks incredibly convenient! This would have applications in other parts of the codebase, too.

I'll see how quick it is to implement in this PR, and if it takes more than an hour or so I'll defer it to another Jira.

BlueSlug replied:
I was unable to get it working, nor was I able to find any sample code (could have sworn I saw some in another project!) or documentation. I have filed [SJRK-451](https://issues.fluidproject.org/browse/SJRK-451) to tackle this update some other time.

cindyli commented:
Clicking the "sign up" link gives:

```
isError: true
message: "Cannot GET /signUp.html"
```

Are there special configs needed for the sign up to work? I'm using "aihec" theme.

cindyli commented:
I didn't find the time out control for the login session. I wonder once a user logs in, will he/she remains logged in until an explicit logout?

jobara commented:
> Clicking the "sign up" link gives:
> 
> ```
> isError: true
> message: "Cannot GET /signUp.html"
> ```
> 
> Are there special configs needed for the sign up to work? I'm using "aihec" theme.

@cindyli I don't believe that Signup has been implemented in this PR yet. At the moment you can use Postman or other tool to send the signup request to `/signup`.

[StoryTelling.postman_collection.json.zip](https://github.com/fluid-project/sjrk-story-telling/files/6022013/StoryTelling.postman_collection.json.zip) contains my collection of Postman requests that I used to test the server. You can use the signup request from there to create an account.

cindyli commented:
> > Clicking the "sign up" link gives:
> > ```
> > isError: true
> > message: "Cannot GET /signUp.html"
> > ```
> > 
> > 
> > Are there special configs needed for the sign up to work? I'm using "aihec" theme.
> 
> @cindyli I don't believe that Signup has been implemented in this PR yet. At the moment you can use Postman or other tool to send the signup request to `/signup`.

Thanks for the information, @jobara. Is the API documentation planned to be added in this PR or later?

jobara committed:
SJRK-403: Fix broken initialization test

BlueSlug committed:
Merge pull request #40 from jobara/SJRK-404

SJRK-403: Fix broken initialization test

It may be the case in the future that we'll want to revive the if/else block for `testCase.createNewStoryOnServer`, but for now since none of the test cases use it, we may as well remove it

jobara commented:
> Thanks for the information, @jobara. Is the API documentation planned to be added in this PR or later?

Which API docs are those? The [end-points](https://github.com/fluid-project/sjrk-story-telling/blob/main/docs/END-POINTS.md) were documented previously, if that's what you are referring to.

cindyli commented:
> > Thanks for the information, @jobara. Is the API documentation planned to be added in this PR or later?
> 
> Which API docs are those? The [end-points](https://github.com/fluid-project/sjrk-story-telling/blob/main/docs/END-POINTS.md) were documented previously, if that's what you are referring to.

Right. They are in the pull request that implements the server side. Now I remember. Thanks for pointing out, @jobara.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -35,6 +35,16 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
+            },
+            "story.createNewStoryOnServer": {
+                record: {
+                    "": {
```
With `createNewStoryOnServer` triggered by the model change to the root path of the `story` component, it means every story content change will also trigger this listener, which might be unnecessary. I wonder if it's possible to use `includeSource: "init"` to only trigger this listener at the component initialization?

BlueSlug replied:
`init` is too early for this to fire, since that is effectively the same as having it call `createNewStoryOnServer` on page load, which is not what we're aiming for at the moment since we want to avoid clogging up the database with story records.

Currently, the function that's called by the invoker checks whether there's an `id` key on the story model, which is a workable solution in the "short" term, but ideally we would have a no-op for changes after the "first" one. It's an interesting challenge.

@jobara and @amb26 had a discussion in Matrix last week (see "Fluid Technology" from 2021-02-17), and @amb26's suggestion of using a flag to track the first model change and attaching a listener to that flag seems like a promising solution to this awkwardness:

> Just create a model listener which listens to "" and produces a {changePath: "userInteractionReceived", value: true} and then attach the listener which checks whether there is an id to that flag

cindyli replied:
I see. Thanks for the explanation, @BlueSlug.

cindyli commented on src/ui/base-page.js:
```diff
@@ -164,6 +171,84 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
     });
 
+    // mix-in grade to include login/logout controls on the page
+    fluid.defaults("sjrk.storyTelling.base.page.withAuthorControls", {
```
This component is more than holding author account information. It's this plus the logout functionality. The current component structure is a bit messy in the way that:
```
sjrk.storyTelling.base.page.withAuthorControls = auth info + logout functionality
sjrk.storyTelling.base.page.login = sjrk.storyTelling.base.page.withAuthorControls + login functionality
```

The login component seems serving the entire account management with both login and logout. What I'm thinking is:
1. Rename the use of `login` for both the component name and the file name `base-page-login.js` to `accountManager`;
2. Move this component into the `accountManager` file;
3. Either create a single `accountManager` component containing login, logout, author account management functionalities, or split 3 functionalities into 3 separate components then assemble.

BlueSlug replied:
I agree that the current scheme is a bit clunky and not entirely accurate, and I was having trouble coming up with an appropriate name for this mix-in grade that didn't seem too much like a boolean variable type name like `hasAccountControls` or `isLoginEnabled`. The `ui` called `authorControls` contains the user greeting, the logout button and the link to the Login and Sign Up pages, hence why this mix-in is a page "with author controls". Those elements should be included on every page for a site that has accounts enabled.

If we implement this proposed setup as I understand it, then we would need to include the code for the login process on any page that we want to have the authorControls included on, which would increase the payload.

The structure I chose was primarily because the login elements only need to be present on the actual login page (under the current design, at least), since it is its own discrete HTML page. I feel that it might be misleading for the page that an author would go to for logging in to (and its underlying `ui` component) be called `accountManager`, as there isn't otherwise any account management happening.

cindyli replied:
Please add this reason that results in the current component design as a code comment. Thanks.

cindyli commented on src/ui/ui-authorControls.js:
```diff
@@ -0,0 +1,41 @@
+(function ($, fluid) {
+
+    // a UI representing the "author controls" part of every page
+    fluid.defaults("sjrk.storyTelling.ui.authorControls", {
```
The same suggestion as above for the UI part of the account management.

cindyli commented:
I ran into an issue with the login page. Here are steps:

1. I followed @jobara's suggestion to manually create an account by sending a request to /signup endpoint;
2. The account appears in the CouchDB so it should have been created successfully;
3. Go to the login page, enter the email and password, hit this error:
<img width="373" alt="Screen Shot 2021-02-23 at 11 56 41 AM" src="https://user-images.githubusercontent.com/572877/108878470-4b2ec080-75ce-11eb-9ec8-5efaa706d93d.png">

Did I do something wrong or miss anything? Thanks.

BlueSlug committed:
SJRK-403: add tests for createNewStoryOnServer function

This update also removes the call to
stringify the server response when
saving stories, and likewise the JSON
parse call on the client side, since
they prevent these tests from working
properly.

BlueSlug committed:
SJRK-403: clarify comments on authorControls ui and mix-in page grade

jobara committed:
SJRK-452: Create new story after block creation or story model changes.

jobara committed:
Merge remote-tracking branch 'BlueSlug/SJRK-404' into SJRK-404

BlueSlug committed:
Merge pull request #41 from jobara/SJRK-404

SJRK-452: Create new story after block creation or story model changes.

BlueSlug committed:
SJRK-402: split login page guts into a grade to share with sign-up page

"authentication" isn't the best name, but it's
the best one we could come up with in short order

BlueSlug committed:
SJRK-402: add base theme signup page and associated grades (WIP)

BlueSlug committed:
SJRK-404: make logout redirect to login page rather than reload current

jobara committed:
SJRK-404: Trigger logout on page load if session no longer valid

BlueSlug committed:
SJRK-402: implement password confirmation comparison validation

BlueSlug committed:
Merge pull request #42 from jobara/SJRK-404

SJRK-404: Trigger logout on page load if session no longer valid

BlueSlug committed:
SJRK-404: merge remote SJRK-404 branch into local after merging PR

BlueSlug committed:
SJRK-403: redirect to site root on login if no previous page is set

BlueSlug commented:
There is currently an issue triggered by the client-side session check (in [`ui-authorControls.js`](https://github.com/BlueSlug/sjrk-story-telling/blob/17ea3f34b6443a30d81c97c011aa160f39d01bec/src/ui/ui-authorControls.js#L36-L42)) which results in the error described by [SJRK-454](https://issues.fluidproject.org/browse/SJRK-454).

BlueSlug committed:
SJRK-454: disable session check due to server crash that ends session

BlueSlug commented:
Considering that whether the model listener that triggers the [SJRK-454](https://issues.fluidproject.org/browse/SJRK-454) bug is enabled or not, the author may still encounter the situation where their session has expired on the server and they cannot publish without logging on again, and considering that I will be pausing my development on this project for an indeterminate amount of time, I have decided to disable the model listener to make the tool usable again.

Ready for more review! :)

cindyli commented on src/ui/base-page-signup.js:
```diff
@@ -0,0 +1,140 @@
+                        confirm: null, // the new author's password confirmation
+                        confirmComparison: null // the intermediate comparison holder
+                    },
+                    modelListeners: {
+                        confirmComparison: {
+                            this: "console",
+                            method: "log",
+                            args: ["comparison from", "{change}.oldValue", "to", "{change}.value"]
```
This debugging code can be removed.

BlueSlug replied:
Whoops, good catch! I usually take care to remove those, but I obviously forgot about this one

cindyli commented on src/ui/base-page-authentication.js:
```diff
@@ -0,0 +1,361 @@
+            // overwrite the listener to prevent losing the previous page URL
+            "onCreate.setCurrentPage": null,
+            "onAuthenticationRequested.resetEmailErrorMessage": {
+                func: "{that}.resetEmailErrorMessage",
```
It will simplify the component definition by moving what's defined in the `resetEmailErrorMessage()` invoker directly to here. It also saves other developers from looking up the invoker level.

The same comment applies to other locations for example line 112, 125, 146 as well as the same usage in other components.

cindyli commented on src/ui/base-page-signup.js:
```diff
@@ -0,0 +1,140 @@
+        },
+        listeners: {
+            "onAuthenticationRequested.resetPasswordConfirmationErrorMessage": {
+                func: "{that}.resetPasswordConfirmationErrorMessage",
```
The comment above about bypassing the extra layer of invokers applies to here and line 50 too.

cindyli commented:
> I ran into an issue with the login page. Here are steps:
> 
>     1. I followed @jobara's suggestion to manually create an account by sending a request to /signup endpoint;
> 
>     2. The account appears in the CouchDB so it should have been created successfully;
> 
>     3. Go to the login page, enter the email and password, hit this error:
> 
> 
> <img alt="Screen Shot 2021-02-23 at 11 56 41 AM" width="373" src="https://user-images.githubusercontent.com/572877/108878470-4b2ec080-75ce-11eb-9ec8-5efaa706d93d.png">
> 
> Did I do something wrong or miss anything? Thanks.

It was my bad that I didn't re-install npm packages. It's gone with a fresh `npm install`.

BlueSlug committed:
SJRK-403: remove superfluous invokers, call functions from listeners

cindyli commented:
Merged at e8fb396d078ee557b6ee58b7d58805ddc9d47fe1
